scipy/sparse/coo.py

```diff
@@ -496,7 +496,11 @@ def _mul_vector(self, other):
         return result
 
     def _mul_multivector(self, other):
-        return np.hstack([self._mul_vector(col).reshape(-1,1) for col in other.T])
+        result = np.zeros((other.shape[1], self.shape[0]),
+                          dtype=upcast_char(self.dtype.char, other.dtype.char))
+        for i, col in enumerate(other.T):
+            coo_matvec(self.nnz, self.row, self.col, self.data, col, result[i])
+        return np.matrix(result.T)
```

Comment:
It looks to me like that behavior needs to be preserved, for backwards compat. There are also some tests for this (although not very comprehensive), for example this in ``test_matvec``:

        # check result type
        assert_(isinstance(M * array([1,2,3]), ndarray))
        assert_(isinstance(M * matrix([1,2,3]).T, matrix))